lib/std/math/gamma.zig

```diff
@@ -0,0 +1,328 @@
+        32670,
+        1925,
+        66,
+        1,
+    };
+    var num: T = 0;
+    var den: T = 0;
+    // split to avoid overflow
+    if (abs < 8) {
+        // big abs would overflow here
+        for (0..numerator.len) |i| {
+            num = num * abs + numerator[numerator.len - 1 - i];
+            den = den * abs + denominator[numerator.len - 1 - i];
+        }
+    } else {
+        // small abs would overflow here
+        for (0..numerator.len) |i| {
+            num = num / abs + numerator[i];
+            den = den / abs + denominator[i];
+        }
+    }
+    return num / den;
+}
+
+// precise sin(pi * x)
+// but not for very small x and integer x, we handle those already
+fn sinpi(comptime T: type, x: T) T {
+    const xmod2 = @mod(x, 2); // [0, 2]
+    const n = (@as(u8, @intFromFloat(4 * xmod2)) + 1) / 2; // {0, 1, 2, 3, 4}
+    const y = xmod2 - 0.5 * @as(T, @floatFromInt(n)); // [-0.25, 0.25]
+    return switch (n) {
+        0, 4 => @sin(std.math.pi * y),
+        1 => @cos(std.math.pi * y),
+        2 => -@sin(std.math.pi * y),
+        3 => -@cos(std.math.pi * y),
+        else => unreachable,
+    };
+}
+
+const expect = std.testing.expect;
+const expectEqual = std.testing.expectEqual;
+const expectApproxEqRel = std.testing.expectApproxEqRel;
+
+test "math.gamma" {
+    inline for (&.{f32, f64}) |T| {
+        const eps = @sqrt(std.math.floatEps(T));
+        try expectApproxEqRel(@as(T, 120), gamma(T, 6), eps);
+        try expectApproxEqRel(@as(T, 362880), gamma(T, 10), eps);
+        try expectApproxEqRel(@as(T, 6402373705728000), gamma(T, 19), eps);
+
+        try expectApproxEqRel(@as(T, 332.7590766955334570), gamma(T, 0.003), eps);
+        try expectApproxEqRel(@as(T, 1.377260301981044573), gamma(T, 0.654), eps);
+        try expectApproxEqRel(@as(T, 1.025393882573518478), gamma(T, 0.959), eps);
+
+        try expectApproxEqRel(@as(T, 7.361898021467681690), gamma(T, 4.16), eps);
+        try expectApproxEqRel(@as(T, 198337.2940287730753), gamma(T, 9.73), eps);
+        try expectApproxEqRel(@as(T, 113718145797241.1666), gamma(T, 17.6), eps);
+
+        try expectApproxEqRel(@as(T, -1.13860211111081424930673), gamma(T, -2.80), eps);
+        try expectApproxEqRel(@as(T, 0.00018573407931875070158), gamma(T, -7.74), eps);
+        try expectApproxEqRel(@as(T, -0.00000001647990903942825), gamma(T, -12.1), eps);
+    }
+}
+
+test "math.gamma.special" {
+    inline for (&.{f32, f64}) |T| {
+        try expect(std.math.isNan(gamma(T, -std.math.nan(T))));
+        try expect(std.math.isNan(gamma(T, std.math.nan(T))));
+        try expect(std.math.isNan(gamma(T, -std.math.inf(T))));
+
+        try expect(std.math.isNan(gamma(T, -4)));
+        try expect(std.math.isNan(gamma(T, -11)));
+        try expect(std.math.isNan(gamma(T, -78)));
+
+        try expectEqual(-std.math.inf(T), gamma(T, -0.0));
+        try expectEqual(std.math.inf(T), gamma(T, 0.0));
+
+        try expectEqual(@as(T, -0.0), gamma(T, -200.5));
```

Comment:
fiy: This does no bit equality test, so -0.0 = +0.0 would both match.

Reply:
Sure, will change to using std.math.isNegativeZero and isPositiveZero